zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/ipfs/IPFSNotebookRepo.java

```diff
@@ -0,0 +1,345 @@
+    }
+    boolean success = ipfs.pinRm(revision.id);
+    if (success) {
+      noteRevisions.remove(revision);
+    } else {
+      LOG.warn("Failed to remove revision " + revision);
+    }
+    saveToFile();
+  }
+
+  /*
+   * Notes are stored in local directory and are added to ipfs only when
+   * committed. If there is no change in note same hash will be generated
+   * and it will not be added to the file.
+   */
+  @Override
+  public Revision checkpoint(String noteId, String commitMessage, AuthenticationInfo subject)
+      throws IOException {
+    Note note = get(noteId, subject);
+    String json = gson.toJson(note);
+    String addResult = ipfs.add(json);
+    if (addResult == null) {
+      throw new IOException("Failed to checkpoint note with ID " + noteId);
+    }
+    Type type = new TypeToken<Map<String, String>>() {
+    }.getType();
+    Map<String, String> data = gson.fromJson(addResult, type);
+    String hash = data.get("Hash");
+
+    /*
+     * The `time` param in Rev is in int(unix timestamp) seconds
+     * System.currentTimeMillis() returns long milliseconds
+     */
+    int time = (int) (System.currentTimeMillis() / 1000L);
+    Revision revision = new Revision(hash, commitMessage, time);
+    List<Revision> noteVersions = noteHashes.get(noteId);
+    if (noteVersions == null) {
+      noteVersions = new ArrayList<>();
+      noteVersions.add(revision);
+      noteHashes.put(noteId, noteVersions);
+    } else {
+      if (!noteVersions.contains(revision)) {
+        noteVersions.add(revision);
+      }
+      /*
+       * revision already exists before. Should i change time ?
+       *
+       * */
+    }
+    saveToFile();
+    LOG.info("Checkpoint for Note " + noteId + "  IpfsRevision is " + revision.id);
+    return revision;
+  }
+
+
+  public Map<String, List<Revision>> getNoteHashes() {
+    return new HashMap<>(noteHashes);
+  }
+
+  /**
+   * get a particular revision from ipfs
+   */
+  @Override
+  public Note get(String noteId, String revId, AuthenticationInfo subject) throws IOException {
+    Note note = null;
+    String pinJsonResponse = ipfs.pinLs();
+    Map<String, Object> data = gson.fromJson(pinJsonResponse, pinType);
+    Map<String, Object> allPinnedObjects = (Map<String, Object>) data.get("Keys");
+
+    if (allPinnedObjects.containsKey(revId)) {
+      note = getNote(revId);
+    }
+    if (note == null) {
+      LOG.warn("revision " + revId + " not present for note " +
+          noteId + " in local ipfs storage");
+    }
+    return note;
+  }
+
+  @Override
+  public List<Revision> revisionHistory(String noteId, AuthenticationInfo subject) {
+    List<Revision> versionHistory = noteHashes.get(noteId);
+    return versionHistory;
+  }
+
+  @Override
+  public void close() {
+    super.close();
+    try {
+      ipfs.getClient().close();
+    } catch (IOException e) {
+      LOG.info("Couldn't successfully close the ipfsClient");
```

Comment:
It's good practice to propagate throwables to logs everywhere, like `LOG.info("Couldn't successfully close the ipfsClient", e)`